ts/tat-trend.ts

```diff
@@ -0,0 +1,365 @@
+        ? new Date(row["Full-Depth (FD) Completed"])
+        : null;
+      days = row["FD Total Days"];
+      break;
+    case "All Completed":
+      const completedDates = [
+        row["CR Analysis Review Completed"],
+        row["CR Release Approval Completed"],
+        row["CR Release Completed"],
+        row["DR Analysis Review Completed"],
+        row["DR Release Approval Completed"],
+        row["DR Release Completed"],
+      ]
+        .map((date) => (date ? new Date(date) : null))
+        .filter((date) => date !== null);
+      completedDate = completedDates.length
+        ? new Date(Math.max(...completedDates.map((date) => date!.getTime())))
+        : null;
+      days =
+        selectedDataType === "CR" ? row["CR Total Days"] : row["DR Total Days"];
+      break;
+    default:
+      completedDate = isCommonGate
+        ? row[`${gate} Completed`]
+          ? new Date(row[`${gate} Completed`])
+          : null
+        : row[`${selectedDataType} ${gate} Completed`]
+        ? new Date(row[`${selectedDataType} ${gate} Completed`])
+        : row[`${gate} Completed`]
+        ? new Date(row[`${gate} Completed`])
+        : null;
+      days = isCommonGate
+        ? row[`${gate} Days`]
+        : row[`${selectedDataType} ${gate} Days`] || row[`${gate} Days`];
+      break;
+  }
+
+  return { completedDate, days };
+}
+
+function getGroupKey(date: Date, selectedGrouping: string): string {
+  const year = date.getFullYear();
+  const quarter = Math.floor(date.getMonth() / 3 + 1);
+
+  switch (selectedGrouping) {
+    case "week":
+      return getWeek(date);
+    case "month":
+      return `${year}-${String(date.getMonth() + 1).padStart(2, "0")}`;
+    case "fiscalQuarter":
+      return `${year} Q${quarter}`;
+    case "fiscalYear":
+      return year.toString();
+    default:
+      return `${year} Q${quarter}`;
+  }
+}
+
+function groupData(
+  jsonData: any[],
+  selectedGrouping: string,
+  selectedGates: string[],
+  selectedDataType: string
+): AssayGroups {
+  const assayGroups: AssayGroups = {};
+
+  jsonData.forEach((row: any) => {
+    const assay = row["Assay"];
+    const totalDays = row[`${selectedDataType} Total Days`];
+    const caseId = row["Case ID"];
+    const releaseCompletedDate = row[`${selectedDataType} Release Completed`];
+
+    if (assay && totalDays != null && caseId != null) {
+      if (!assayGroups[assay]) {
+        assayGroups[assay] = {};
+      }
+
+      selectedGates.forEach((gate) => {
+        const { completedDate, days } = getCompletedDateAndDays(
+          row,
+          gate,
+          selectedDataType
+        );
+
+        if ((completedDate && days != null) || gate === "All Completed") {
+          const date = new Date(
+            gate === "All Completed"
+              ? releaseCompletedDate || new Date()
+              : completedDate
+          );
+          const groupKey = getGroupKey(date, selectedGrouping);
```

Comment:
This is grouping by the completion date of the individual QC gate/step, and I'm not sure that's what we want. The other option is to group all steps by case completion date, which is what I am thinking we want, but I'm not sure.

The advantage to always using the case completion date is that if you filter by case completion date (e.g. cases completed in 2023), then all data will be in that range. With this implementation, you can filter to include cases completed in 2023 and some of your data will be outside of that - cases started in 2022 and finished in 2023 for example.

The current implementation includes cases that are incomplete, and I'm not sure whether we want that either.

Should discuss this more with QA

Reply:
going with  group all steps by case completion date for now, will check with QA